crates/fuel-core/src/graphql_api/validation_extension.rs

```diff
@@ -0,0 +1,97 @@
+use async_graphql::{
+    extensions::{
+        Extension,
+        ExtensionContext,
+        ExtensionFactory,
+        NextResolve,
+        ResolveInfo,
+    },
+    Pos,
+    ServerError,
+    ServerResult,
+    Value,
+};
+use std::{
+    collections::{
+        hash_map::Entry,
+        HashMap,
+    },
+    sync::Arc,
+};
+
+/// The extension validates the queries.
+pub(crate) struct ValidationExtension {
+    recursion_limit: usize,
+}
+
+impl ValidationExtension {
+    pub fn new(recursion_limit: usize) -> Self {
+        Self { recursion_limit }
+    }
+}
+
+impl ExtensionFactory for ValidationExtension {
+    fn create(&self) -> Arc<dyn Extension> {
+        Arc::new(ValidationInner::new(self.recursion_limit))
+    }
+}
+
+struct ValidationInner {
+    recursion_limit: usize,
+    visited: parking_lot::Mutex<HashMap<String, usize>>,
```

Comment:
Is this so performance critical that we need the parking lot mutex instead of the std one? If we introduce parking lot, I assume we'd likely want to follow up with using their synchronization primitives everywhere (which could be a separate issue).